init: Coalesce Chainstate loading sequence between {,non-}unittest codepaths

dongcarl opened the pull request:
This PR:
1. Coalesce the Chainstate loading sequence between `AppInitMain` and `*TestingSetup` (which makes it more tested)
2. Makes the Chainstate loading sequence reusable in preparation for future work extracting out our consensus engine.

Code-wise, this PR:
1. Extracts `AppInitMain`'s Chainstate loading sequence into a `::LoadChainstateSequence` function
2. Makes this `::LoadChainstateSequence` function reusable by
    1. Decoupling it from various concepts (`ArgsManager`, `uiInterface`, etc)
    2. Making it report errors using an `enum` rather than by setting a `bilingual_str`
3. Makes `*TestingSetup` use this new `::LoadChainstateSequence`

Reviewers: Aside from commentary, I've also included `git diff` flags of interest in the commit messages which I hope will aid review!

dongcarl committed:
node/chainstate: Decouple from ArgsManager

...instead pass in only the necessary information

dongcarl committed:
node/chainstate: Decouple from stringy errors

This allows us to separate the initialization code from translations and
error reporting.

This change changes the caller semantics of LoadChainstate quite
drastically.

To see that this change doesn't change behaviour, observe that:

1. Prior to this change, LoadChainstate returned false only in the "bad
   genesis block" failure case (by returning InitError()), indicating
   that the caller should immediately bail. After this change, the
   corresponding ERROR_BAD_GENESIS_BLOCK handler in src/init.cpp
   maintains behavioue by also bailing immediately.

2. The failed_* temporary booleans were only used to break out of the
   outer do/while(false) loop. They can therefore be safely removed.

dongcarl committed:
node/chainstate: Decouple from concept of uiInterface

...instead allow the caller to optionally pass in callbacks which are
triggered for certain events.

Behaviour change: The string "Verifying blocks..." was previously
printed for each chainstate in chainman which did not have an
effectively empty coinsview, now it will be printed once unconditionally
before we call VerifyLoadedChain.

dongcarl committed:
node/chainstate: Reduce coupling of LogPrintf

...by moving the try/catch out of LoadChainstate

I strongly recommend reviewing with the following git-diff flags:
  --color-moved=dimmed_zebra --color-moved-ws=allow-indentation-change

dongcarl committed:
init: Delay RPC block notif until warmup finished

See added code comment for more details.

dongcarl committed:
validation: VerifyDB only needs Consensus::Params

Previously we were passing in CChainParams, when VerifyDB only needed
the Consensus::Params subset.

dongcarl committed:
node: Extract chainstate loading sequence

I strongly recommend reviewing with the following git-diff flags:
  --color-moved=dimmed_zebra --color-moved-ws=allow-indentation-change

[META] This commit is intended to be as close to a move-only commit as
       possible, and lingering ugliness will be resolved in subsequent
       commits.

A few variables that are passed in by value instead of by reference
deserve explanation:

- fReset and fReindexChainstate are both local variables in AppInitMain
  and are not modified in the sequence

- fPruneMode, despite being a global, is only modified in
  AppInitParameterInteraction, long before LoadChainstate is called

----

[META] This semantic will change in a future commit named
       "node/chainstate: Decouple from stringy errors"

dongcarl committed:
node/chainstate: Decouple from concept of NodeContext

...instead pass in only the necessary information

Also allow mempool to be a nullptr

dongcarl committed:
node/chainstate: Decouple from GetTimeMillis

...instead just move it out

dongcarl committed:
node/chainstate: Remove do/while loop

I strongly recommend reviewing with the following git-diff flags:
  --ignore-space-change

dongcarl committed:
node/caches: Extract cache calculation logic

I strongly recommend reviewing with the following git-diff flags:
  --color-moved=dimmed_zebra --color-moved-ws=allow-indentation-change

[META] In a future commit, this function will be re-used in TestingSetup
       so that the behaviour matches across test and non-test init
       codepaths.

dongcarl committed:
node/chainstate: Add options for in-memory DBs

[META] In a future commit, these options will be used in TestingSetup to
       ensure that the DBs are in-memory.

dongcarl committed:
test/setup: Use LoadChainstate

This commit coalesces the chainstate loading sequence between our unit
test and non-unit test init codepaths.

dongcarl committed:
node/chainstate: Decouple from GetTime

...instead pass in a std::function<int64_t()>

Note that the static_cast is needed (apparently) for the compiler to
know which overloaded GetTime to choose.

dongcarl committed:
node/chainstate: Decouple from ShutdownRequested

...instead allow optionally passing in a std::function<bool()>

jamesob commented:
Concept ACK! I've already done a preliminary review of this code but will formally test and review very soon.

jamesob commented on src/init/caches.cpp:
```diff
@@ -0,0 +1,36 @@
+#include <util/system.h>
+#include <validation.h>
+
+void CalculateCacheSizes(const ArgsManager& args, CacheSizes& cache_sizes, size_t n_indexes)
```
1f966b8d9f

One naive question here might be: why are we using a mutable arg for `cache_sizes` vs. just returning a constructed `CacheSizes` instance to make the function pure? Answer seems to be that this is used within a unittest context to modify an existing `m_cache_sizes` instance.

dongcarl replied:
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

jamesob reviewed (approved):
ACK 7dd898c0df678859ca92773b6990e4e28a83ab24 ([`jamesob/ackr/23280.2.dongcarl.init_coalesce_chainstate`](https://github.com/jamesob/bitcoin/tree/ackr/23280.2.dongcarl.init_coalesce_chainstate))

This is an excellent change in terms of cleaning up and modularizing our messy (and previously un-unittested) initialization code. I had actually taken a run at doing this previously as a part of #15606 but gave up because I thought it'd be too big a change. Carl has done a much better job of implementing it here than I had anyway, but for what it's worth the design I had sketched is very similar to this one.

This patch paves the way for further libbitcoinkernel work - but even aside from that, it ensures that we're reusing more (all?) of our chain initialization logic from within the unittests which has quite a bit of value on its own.

I wrote #23289 specifically to test this change, and I've run it ~25 times on HEAD. I've also tested this branch with a pre-existing mainnet datadir, booting and interrupting init a number of times.

Great job, @dongcarl!

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 7dd898c0df678859ca92773b6990e4e28a83ab24 ([`jamesob/ackr/23280.2.dongcarl.init_coalesce_chainstate`](https://github.com/jamesob/bitcoin/tree/ackr/23280.2.dongcarl.init_coalesce_chainstate))

-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmF8OJcACgkQepNdrbLE
TwW1uxAAjTIPIZ4ddUhSlAax8GQR43wEE3Qvk5+I51RHNGyIS8QoPb0lgmBYgqJW
bqvVyMC0cKg9QNDLELrQQ/ySVf2Vl9H/6va4P18ZAKLg8cNuyeVr3cl4Swe0ZH9F
clRa8b75NTRUtOZlt34O2eTfigyKbnQ5L1+G8HJWNoBvWEPt5JSV7G2UU1tyLEh3
3lvS5vHivSt+8WqIhB/lBIERJmI7vHjFAhN8P6SgLyQr7NxKA8Lw6cFUFeJIWKDP
pkza4uabu1cDqAurxO2jgIq4GDL7FXz3fkXcFzGTPgY1JlXI26dEkuWBSEA42zGs
ubizjy/JIyRUbzRv5jBia5woDJgR40eFfsb7GkZOHfZtoQsDN85xwvZpRCzl/YnY
y1exK8fGSmCuP6cIP2mbbz9TrUlrGfyZm0vwIiqrRpGtO7q1CdTyiV7gnOwRUCjY
LqUI2WXMuJGSD/WaPIghpllFBejadETVkfjMUHBtaTO8WdpPchdgnLZFtKlyjvLV
x1UkNk3Xqkzzk75+XPEK+GBSDPnaD9fDnfypu7f18pin2O6gGFVErQto67xM1+HY
0eRlQOhM/T+SdsDWcl9z4fUS1pA+zxxpr3BRNHR4aNbuxXj326VIVnQUatEotncy
ilPh7NYFFVMWpoOGG5HbucjlazQLf/Z4Q9FxUqdDglXRSqK+kQM=
=XedH
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-5.10.0-9-amd64-x86_64-with-glibc2.28

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis --enable-wallet --enable-debug --with-daemon --enable-natpmp-default CXX=/usr/bin/clang++ CC=/usr/bin/clang PYTHONPATH= --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --disable-openssl-tests --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --no-create --no-recursion

Compiled with /usr/bin/ccache /usr/bin/clang++ -std=c++17 -mavx -mavx2 -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_top_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4 -msha -msse4.1 -msse4.2  i

Compiler version: Debian clang version 13.0.1-++20211019122913+8a93745a7121-1~exp1~20211019003538.10
```

</p></details>

dongcarl commented:
My sincerest thanks for the thorough review @jamesob, that functional test is certainly beyond the call of duty, and the `master` bug it revealed is leading to some excellent investigation in #23365. 🎉 

I'm glad to hear that this sort of cleanup came up for the AssumeUTXO work as well, my hope is that the change will make future work easier involving the init sequence easier to reason about.

> reusing more (all?) of our chain initialization logic from within the unittests

One thing to note is that although we _load_ the chainstate, we do not _activate_ it, which is done in `ThreadImport` with `ActivateBestChain`, this can be included in followup PRs.

jnewbery commented on src/init.cpp:
```diff
@@ -24,6 +24,8 @@
 #include <index/coinstatsindex.h>
 #include <index/txindex.h>
+#include <init/caches.h> // for CalculateCacheSizes
+#include <init/chainstate.h> // for LoadChainstateSequence
```
I agree. These comments are unmaintainable.

jnewbery commented on src/init/caches.h:
```diff
@@ -0,0 +1,22 @@
+    int64_t tx_index_cache_size;
+    int64_t filter_index_cache_size;
+};
+void CalculateCacheSizes(const ArgsManager& args, CacheSizes& cache_sizes, size_t n_indexes = 0);
```
Again, it'd be nice to have a doxygen comment here. It's not immediately obvious what `n_indexes` represents.

Is there any reason that you made `cache_sizes` an out-param rather than the return value for this function?

Since this is a new function and you're adding all the call sites, can we make `n_indexes` a required argument?

jamesob replied:
> Is there any reason that you made cache_sizes an out-param rather than the return value for this function?

https://github.com/bitcoin/bitcoin/pull/23280#discussion_r739436153

dongcarl replied:
In my latest push (6736e24a490d756e0e1e70b9e212775cf305ebe6), I ended up having `CalculateCacheSizes` return a CacheSize, which is perhaps a bit more readable.

jnewbery commented on src/init/caches.h:
```diff
@@ -0,0 +1,22 @@
+class ArgsManager;
+
+struct CacheSizes {
+    int64_t block_tree_db_cache_size;
```
The `_cache_size` parts to all these member names seems redundant with the fact they're members of a struct called `CacheSizes`

dongcarl replied:
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

jnewbery commented on src/init.cpp:
```diff
@@ -1361,217 +1363,100 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+                                        cache_sizes.coin_cache_usage_size,
+                                        args.GetIntArg("-checkblocks", DEFAULT_CHECKBLOCKS),
+                                        args.GetIntArg("-checklevel", DEFAULT_CHECKLEVEL),
+                                        false,
+                                        false,
```
Can you add inline comments for these arguments (that can be checked by a clang-tidy):

```suggestion
                                        /*block_tree_db_in_memory=*/false,
                                        /*coins_db_in_memory=*/false,
```

jnewbery reviewed (commented):
Concept ACK.

A few questions/observations:

1
Several of the commit logs reference "ACCS" or "ACSS". Sorry if I'm being dense, but what are those?

2
The commit log for _init/chainstate: Decouple from GetTimeMillis_ reads "...instead require caller to pass in a std::function that returns the current system time in milliseconds as a int64_t", but that doesn't match what the code does. Perhaps this is left over from an old version of this branch?

3
Can you explain the benefit of commit _init/chainstate: Reduce coupling of LogPrintf_? It doesn't seem a problem to me for `LoadChainstateSequence()` to call the global logger. Lots of other parts of the codebase do that already. Having the try/catch inside `LoadChainstateSequence()` means that it's noexcept and the caller doesn't need to worry about catching exceptions.

4
In _validation: Call NotifyBlockTip in CCS::LoadChainTip_, you're moving the RPC notification into `CChainstate::LoadChainTip()`. However, that function is also called by `ChainstateManager::ActivateSnapshot()`, so that's potentially a behavior change there. You also change the call from `RPCNotifyBlockChange()` to `uiInterface.NotifyBlockTip()`. However, `uiInterface.NotifyBlockTip` has multiple other "slots", e.g. `BlockNotifyGenesisWait`. Are we sure that isn't also a behavior change?

I think that the `RPCNotifyBlockChange()` call can just be brought down to immediately before RPC leaves warmup:

https://github.com/bitcoin/bitcoin/blob/77a2f5d30c5ecb764b8a7c098492e1f5cdec90f0/src/init.cpp#L1842

There can't be any pending RPC calls waiting for the blockchange notification (since RPC is still in warmup) so I think the only purpose of this call is to set the `latestblock` fields.

5
In _init/chainstate: Decouple from GetAdjustedTime_, why not just pass in the time instead of a function that returns the time?

jnewbery commented on src/init/chainstate.cpp:
```diff
@@ -0,0 +1,157 @@
+    };
+
+    {
+        LOCK(cs_main);
```
This function now consists of three code blocks that each lock `cs_main`. What do you think about just taking `cs_main` at the top of the function and holding it throughout?

dongcarl replied:
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

ryanofsky reviewed (commented):
This looks very promising! Mostly reviewed this

- [X] 83ed7db27b4c4798666b77d83fc8e7c3eab8c621 init: Extract chainstate loading sequence (1/17)
- [X] b094ca4fd89b50c7179355f72113d794725f8a9d init/chainstate: Decouple from ArgsManager (2/17)
- [X] 3ecaf812d4934d63124b640d2aeb008d4b742d89 init/chainstate: Decouple from concept of NodeContext (3/17)
- [X] e53e85f8b4ff7642c876950942cadcc819aa50e1 init/chainstate: Decouple from GetTimeMillis (4/17)
- [X] cd29d334a38fdaf0bbbe810a9cbdeeffad394daa init/chainstate: Decouple from stringy errors (5/17)
- [X] 25cea87e6ea4a7cf59f38df1454265b2de8461c9 init/chainstate: Remove do/while loop (6/17)
- [X] 0e61af06834c9a9331e620343e60e9104e08331e init/chainstate: Decouple from concept of uiInterface (7/17)
- [X] 38d88e2950afc99caca31388fd75f03c4d06bcc9 init/chainstate: Reduce coupling of LogPrintf (8/17)
- [X] 10564583429762ddfb6601d1f64ac663406bf8e2 init/chainstate: Move -checkblocks limitation to help (9/17)
- [ ] b229024932ebd4775070149ff0f27d0fd329bf0a validation: Call NotifyBlockTip in CCS::LoadChainTip (10/17)
- [X] 8eae069b0015524d877a6fc9f92ab85d2ad6b26b init/chainstate: Decouple from GetAdjustedTime (11/17)
- [X] 855a86102095003779d63b514728d1f0ed4d3a91 init/chainstate: Decouple from ShutdownRequested (12/17)
- [X] 391fe919a8321ed8ff9af990797b2cc737968b78 validation: VerifyDB only needs Consensus::Params (13/17)
- [ ] 1f966b8d9fa41073a27a475715cf984b227197de init/caches: Extract cache calculation logic (14/17)
- [X] f45b885db6636e135d617d915e0ef4ec9e1be8e6 init/chainstate: Add options for in-memory DBs (15/17)
- [ ] 8e2817a66fb9c00b4392256c700c35380d7f7eb3 test/setup: Use LoadChainstateSequence (16/17)
- [ ] 7dd898c0df678859ca92773b6990e4e28a83ab24 test/setup: Unify m_args and gArgs (17/17)

theuni commented on src/init/chainstate.cpp:
```diff
@@ -137,7 +137,7 @@ std::optional<ChainstateLoadingError> LoadChainstateSequence(bool fReset,
 
                 const CBlockIndex* tip = chainstate->m_chain.Tip();
-                if (tip && tip->nTime > GetAdjustedTime() + 2 * 60 * 60) {
+                if (tip && tip->nTime > get_unix_time_seconds() + 2 * 60 * 60) {
```
In commit "init/chainstate: Decouple from GetAdjustedTime"

I suppose we can't just pass a timeout time in here because the current logic includes the runtime of the loading/checking itself?

This seems like yet something else that would be unnecessary if this function were split. That keeps coming up as a potential alternative, but I'm unsure if there's something that would prevent it, so I'll stop suggesting it.

theuni commented on src/init/chainstate.cpp:
```diff
@@ -21,6 +20,7 @@ std::optional<ChainstateLoadingError> LoadChainstateSequence(bool fReset,
                                                              unsigned int check_blocks,
                                                              unsigned int check_level,
                                                              std::function<int64_t()> get_unix_time_seconds,
+                                                             std::optional<std::function<bool()>> shutdown_requested,
```
In commit "init/chainstate: Decouple from ShutdownRequested"

Without actually trying it.. I think it would be more straightforward here to use a `CThreadInterrupt` which gets triggered for this.

theuni reviewed (commented):
I like this. `LoadChainstateSequence` is still quite wonky, but at least it's better defined. Nice to have that really weird loop removed too.

I'm curious to see what you think about splitting it up. I only reviewed "Decouple from stringy errors" and "Extract cache calculation logic" lightly. They look good but I'd want to scrutinize those two for edge-cases before I could be confident that they're no-ops.

jamesob commented on src/init.cpp:
```diff
@@ -1361,217 +1363,100 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        }
+        if (rv.has_value()) {
+            switch (rv.value()) {
+            case ChainstateLoadingError::ERROR_LOADING_BLOCK_DB:
```
I agree with @theuni but in any case this feels like bikeshedding to me. What's written as-is is a clear improvement and refining this further should be optional or follow-up IMO.

jnewbery replied:
>  How about a lookup function rather than marrying them? Something like master/src/script/script_error.cpp#L10

I agree that this would be an improvement, separating the error code:string lookup from the recovery logic.

ryanofsky replied:
In commit "init/chainstate: Decouple from stringy errors" (cd29d334a38fdaf0bbbe810a9cbdeeffad394daa)

re: https://github.com/bitcoin/bitcoin/pull/23280#discussion_r743709991

I agree with John's original comment, and do not think fine grained error codes are good or that some kind of external error code -> string indirection would be good. I think the ideal thing would be for this to return a straightforward SUCCESS / FAILURE / INTERRUPTED status, with a detailed error string or list of error strings in the case of failure. The problem with fine grained return codes is:

- They are maintenance burden making code more verbose than it needs to be and hard to spot inconsistencies easier to introduce.
- They are unergnomic for callers (important if this is supposed evolve into a libbitcoin kernel API) because it's more work for callers to figure out how to handle a long list error codes correctly than a small set of cases.
- They are worse for user experience because it's not possible to include specific information about errors that may be helpful like filenames, timestamps, hashes.

dongcarl committed:
Split off VerifyLoadedChainstate

dongcarl commented on src/init/chainstate.h:
```diff
@@ -0,0 +1,30 @@
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_INIT_CHAINSTATE_H
```
Ah I see! What do you think about putting the files under `src/node/init`? And I supposed everything would be under the `node::init` namespace?

ryanofsky replied:
> Ah I see! What do you think about putting the files under `src/node/init`? And I supposed everything would be under the `node::init` namespace?

Probably best to use your own judgement, and other people may have opinions, but IMO just moving things into top level, `node`, `wallet`, `util` directories is ideal for seeing where things come from and making it obvious when you may be doing something bad like calling wallet code from node code. Extra nesting may be good, but I think only if there is some specific justification for it. In this case, I think having an "init" module is not good, because it will encourage people to add startup code from disparate parts of code there instead of closer to the actual database/index/connection structures that are being initialized. 

I also think it's nice when there's a 1:1 correspondence between directories and namespaces and personally would not suggest extra nesting for namespaces either.

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/23280#discussion_r743133511

FWIW, followed up with more `node::` namespace in #/23497

dongcarl committed:
Remove all #include // for * comments

dongcarl committed:
Move init logistics message for BAD_GENESIS_BLOCK to init.cpp

dongcarl committed:
Move mempool nullptr Assert out of LoadChainstate

ryanofsky commented on src/init/chainstate.cpp:
```diff
@@ -0,0 +1,206 @@
+bool LoadChainstateSequence(bool& fLoaded,
+                            bilingual_str& strLoadError,
+                            bool fReset,
+                            CClientUIInterface& uiInterface,
```
re: https://github.com/bitcoin/bitcoin/pull/23280#discussion_r743923908

> In commit "init: Extract chainstate loading sequence"
> 
> Could you please use a new variable here rather than shadowing the global? I was very confused about what changed that required `AddReadErrCallback([](...))` -> `AddReadErrCallback([&](...))`

IMO ideal thing to do here so this commit remains moveonly and easy to review and doesn't require thinking about renames or new variables or adding `[&]`, is just to keep `uiInterface` a global instead of a parameter here, since it is already going away later anyway in 83ed7db27b4c4798666b77d83fc8e7c3eab8c621.

dongcarl replied:
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

ryanofsky reviewed (commented):
Just posting few more comments. I do like Cory's Load/Verify split idea, too.

dongcarl committed:
Move -checkblocks LogPrintf to AppInitMain

dongcarl committed:
Collapse the 2 cs_main locks in LoadChainstate

dongcarl commented on src/init.cpp:
```diff
@@ -1361,217 +1363,100 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+                                            uiInterface.InitMessage(_("Verifying blocks…").translated);
+                                        });
+        } catch (const std::exception& e) {
+            LogPrintf("%s\n", e.what()); // XXX
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/init/chainstate.cpp:
```diff
@@ -0,0 +1,157 @@
+                                                             std::function<int64_t()> get_unix_time_seconds,
+                                                             std::optional<std::function<bool()>> shutdown_requested,
+                                                             std::optional<std::function<void()>> coins_error_cb,
+                                                             std::optional<std::function<void()>> verifying_blocks_cb) {
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/Makefile.am:
```diff
@@ -547,6 +549,8 @@ libbitcoin_common_a_SOURCES = \
   deploymentinfo.cpp \
   external_signer.cpp \
+  init/caches.cpp \
+  init/chainstate.cpp \
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/init.cpp:
```diff
@@ -1361,217 +1363,100 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
-            } catch (const std::exception& e) {
-                LogPrintf("%s\n", e.what());
+            case ChainstateLoadingError::ERROR_BAD_GENESIS_BLOCK:
+                return false;  // bail immediately!
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6, just copied over the existing comment which mostly describes error handling anyway

dongcarl commented on src/init.cpp:
```diff
@@ -1361,217 +1363,100 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
-                LogPrintf(" block index %15dms\n", GetTimeMillis() - load_block_index_start_time);
-            }
-        } while(false);
+        } else if (!ShutdownRequested()) {
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/init.cpp:
```diff
@@ -1361,217 +1363,100 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
-                // (we're likely using a testnet datadir, or the other way around).
-                if (!chainman.BlockIndex().empty() &&
-                        !chainman.m_blockman.LookupBlockIndex(chainparams.GetConsensus().hashGenesisBlock)) {
-                    return InitError(_("Incorrect or no genesis block found. Wrong datadir for network?"));
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/init.cpp:
```diff
@@ -1415,8 +1412,46 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+                strLoadError = _("Corrupted block database detected");
+                break;
+            }
+        } else if (!ShutdownRequested()) {
```
6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/init.cpp:
```diff
@@ -1415,8 +1412,46 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+                strLoadError = _("Error loading block database");
+                break;
+            case ChainstateLoadingError::ERROR_BAD_GENESIS_BLOCK:
+                return false;  // bail immediately!
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/init/chainstate.h:
```diff
@@ -0,0 +1,30 @@
+class ChainstateManager;
+struct NodeContext;
+
+bool LoadChainstateSequence(bool& fLoaded,
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/init/chainstate.cpp:
```diff
@@ -37,11 +36,11 @@ bool LoadChainstateSequence(bool& fLoaded,
         try {
             LOCK(cs_main);
-            chainman.InitializeChainstate(Assert(node.mempool.get()));
+            chainman.InitializeChainstate(mempool);
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/init/chainstate.cpp:
```diff
@@ -127,36 +115,25 @@ bool LoadChainstateSequence(bool& fLoaded,
-            LogPrintf("%s\n", e.what());
-            strLoadError = _("Error opening block database");
-            break;
+            LogPrintf("%s\n", e.what()); // XXX
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/init/chainstate.cpp:
```diff
@@ -90,11 +90,9 @@ std::optional<ChainstateLoadingError> LoadChainstateSequence(bool fReset,
-                    "", CClientUIInterface::MSG_ERROR);
-            });
+            if (coins_error_cb.has_value()) {
+                chainstate->CoinsErrorCatcher().AddReadErrCallback(coins_error_cb.value());
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/init/chainstate.cpp:
```diff
@@ -138,7 +136,9 @@ std::optional<ChainstateLoadingError> LoadChainstateSequence(bool fReset,
         for (CChainState* chainstate : chainman.GetAll()) {
             if (!is_coinsview_empty(chainstate)) {
-                uiInterface.InitMessage(_("Verifying blocks…").translated);
+                if (verifying_blocks_cb.has_value()) {
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/init.cpp:
```diff
@@ -1400,25 +1400,31 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
-                                             uiInterface.InitMessage(_("Verifying blocks…").translated);
-                                         });
+        std::optional<ChainstateLoadingError> rv;
+        try {
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented on src/init/chainstate.cpp:
```diff
@@ -136,10 +136,6 @@ std::optional<ChainstateLoadingError> LoadChainstateSequence(bool fReset,
                 if (verifying_blocks_cb.has_value()) {
                     verifying_blocks_cb.value()();
                 }
-                if (fHavePruned && check_blocks > MIN_BLOCKS_TO_KEEP) {
```
Fixed as of 6736e24a490d756e0e1e70b9e212775cf305ebe6

dongcarl commented:
Thanks all for the reviews! I addressed a subset of the concerns in my latest 2 pushes. One thing I'd like some help with:

For the RPC block change notification, I decided to go with jnewbery's method here:
     
> In _validation: Call NotifyBlockTip in CCS::LoadChainTip_, you're moving the RPC notification into `CChainstate::LoadChainTip()`. However, that function is also called by `ChainstateManager::ActivateSnapshot()`, so that's potentially a behavior change there. You also change the call from `RPCNotifyBlockChange()` to `uiInterface.NotifyBlockTip()`. However, `uiInterface.NotifyBlockTip` has multiple other "slots", e.g. `BlockNotifyGenesisWait`. Are we sure that isn't also a behavior change?
> 
> I think that the `RPCNotifyBlockChange()` call can just be brought down to immediately before RPC leaves warmup:
> 
> https://github.com/bitcoin/bitcoin/blob/77a2f5d30c5ecb764b8a7c098492e1f5cdec90f0/src/init.cpp#L1842
> 
> There can't be any pending RPC calls waiting for the blockchange notification (since RPC is still in warmup) so I think the only purpose of this call is to set the `latestblock` fields.

This change is in this commit: dea5766ccaed1db77f8ca335d913ecf1ea4d41d1. This seems quite safe to me, and answers some of the questions ryanofsky posed in https://github.com/bitcoin/bitcoin/pull/23280#discussion_r743895731. However, I'm not well-versed in RPC and how that interacts with our multiple chainstates. So I'd love some input on that.

jamesob commented:
reACK https://github.com/bitcoin/bitcoin/pull/23280/commits/2c8f0e948d1ffe2fc729ce51fa42544616b51b5b

I re-reviewed this from scratch (no easier way to do it) after rebasing on master to avoid the peers.dat v3/v4 startup error. Notable changes were `LoadChainstate`/`VerifyLoadedChainstate` split, import comment removal, and `cs::main` lock consolidation, which I think are all fine.

Rerunning my proposed functional tests (#23289) works for the most part, but there is a failure at the end when we intentionally create an init failure by removing block index *.ldb files (https://github.com/bitcoin/bitcoin/pull/23289/commits/d9803f7a0a33688f7429cf10384244f4770851ca#diff-1eb9875f38625ed8a456749c0a96ef72f7a5dc666c7c486f0b585d65680035a0R146). So I guess this change doesn't maintain strict output backwards compatibility, but personally I don't think that matters so much if the issue really is just what messages are shown. 

For what it's worth, the failure is due to this change: https://github.com/bitcoin/bitcoin/pull/23280/commits/18be9fec4967327158ed05f4c3e2531178ba3f86#diff-1767ee60ee6478aaca24348d0fdc0ccdfcf82c35b4543823d9ef39950b50be14L161

```
 node0 2021-11-18T00:06:03.158933Z [init] [dbwrapper.cpp:137] [CDBWrapper] Opening LevelDB in /tmp/bitcoin_func_test_y5h3tzr2/node0/regtest/blocks/index
 node0 2021-11-18T00:06:03.159032Z [init] [dbwrapper.cpp:253] [HandleError] Fatal LevelDB error: Corruption: 1 missing files; e.g.: /tmp/bitcoin_func_test_y5h3tzr2/node0/regtest/blocks/index/000035.ldb
 node0 2021-11-18T00:06:03.159039Z [init] [dbwrapper.cpp:254] [HandleError] You can use -debug=leveldb to get more complete diagnostic messages
 node0 2021-11-18T00:06:03.159101Z [init] [init.cpp:1420] [AppInitMain] Fatal LevelDB error: Corruption: 1 missing files; e.g.: /tmp/bitcoin_func_test_y5h3tzr2/node0/regtest/blocks/index/000035.ldb
 node0 2021-11-18T00:06:03.159121Z [init] [noui.cpp:56] [noui_InitMessage] init message: Verifying blocks…
 test  2021-11-18T00:06:03.190000Z TestFramework.node0 (DEBUG): [node 0] bitcoind exited with status -11 during initialization
 test  2021-11-18T00:06:03.191000Z TestFramework (ERROR): Assertion failed
                                   Traceback (most recent call last):
                                     File "/home/james/src/bitcoin/test/functional/test_framework/test_framework.py", line 132, in main
                                       self.run_test()
                                     File "/home/james/src/bitcoin/./test/functional/feature_init.py", line 163, in run_test
                                       node.assert_start_raises_init_error(
                                     File "/home/james/src/bitcoin/test/functional/test_framework/test_node.py", line 523, in assert_start_raises_init_error
                                       self._raise_assertion_error(
                                     File "/home/james/src/bitcoin/test/functional/test_framework/test_node.py", line 167, in _raise_assertion_error
                                       raise AssertionError(self._node_msg(msg))
                                   AssertionError: [node 0] Expected message "Error opening block database." does not partially match stderr:
                                   ""
 ```

### Process note

I'd request that we scope future reviews to matters of correctness (vs. style) and leave the latter for follow-up in smaller PRs. ACK invalidation on this PR is painful, and as-written it's a very substantial improvement (e.g. unittest init reuse) over what we have now.

theuni commented:
> reACK [2c8f0e9](https://github.com/bitcoin/bitcoin/commit/2c8f0e948d1ffe2fc729ce51fa42544616b51b5b)
> 
> I re-reviewed this from scratch (no easier way to do it)

Blah, but thanks for confirming the best approach :)

Thanks @jamesob for another thorough review. I'll re-review from scratch soon as well.

ryanofsky commented:
The main thing I'm waiting for is for the new code to be moved out of `src/init/` which is for generic process initialization to `src/node/` which is for libbitcoin_server node code. Should be easy to do this with a plain git rebase, or in the worst case with `git filter-branch mv` (https://github.com/bitcoin/bitcoin/pull/23280#discussion_r743858865) if git rename detection chokes.

It would also be nice if LoadChainstate documentation was moved out of commit messages into a code comment for the LoadChainstate declaration so the PR would be easier to understand and review

dongcarl commented:
Pushed 2c8f0e948d1ffe2fc729ce51fa42544616b51b5b -> fdc07a1d2916a85add9b11130a8c2cf6c02a09eb
1. Moved some `LoadChainstate` documentation from commit messages to actual code as suggested by ryanofsky (actually a great idea and made things clearer)
2. Moved the newly added files to `node/` instead of `init/`, fixed up alphabetical ordering where appropriate
3. Added some comments about the nuance of the SHUTDOWN_REQUESTED/PROBED enum return value

w/re point number 3, although I've implemented the reviewers' suggestion (returning an enum variant), I think this solution may actually be a bit more complicated than my original one. This is mostly because we can only guarantee that we detect shutdown requests up to our last call to `ShutdownRequested`, and there may be indirect shutdown requests after that that we won't catch. So in order to maintain behavior / be correct we still need to check `ShutdownRequested()` after we exit the sequence, and the enum return value is rendered somewhat useless: https://github.com/bitcoin/bitcoin/pull/23280/commits/64195bf2e4b8c6f74d0df7ba85a1ee0f0b6c4e5f#diff-b1e19192258d83199d8adaa5ac31f067af98f63554bfdd679bd8e8073815e69dR1450-R1451

ryanofsky commented:
Thanks for updates! Will look more, but I don't think the fact that code might need to call `ShutdownRequested()` later makes having distinct `SUCCESS` / `FAILURE` / `INTERRUPTED` return codes useless. If calling code wants to treat `SUCCESS` and `INTERRUPTED` values the same way (by not printing errors, and going on to check whether a shutdown is currently requested), that's great, but it doesn't suggest to me that `SUCCESS` and `INTERRUPTED` should be merged together. `SUCCESS` and `INTERRUPTED` do actually mean different things, and it also would be reasonable for calling code to treat them differently, even if it isn't now, for example by logging the interruption when `INTERRUPTED` is returned, or by returning `INTERRUPTED` to the caller of the caller, and moving Shutdown handling to a higher level.

jamesob commented:
ACK fdc07a1d2916a85add9b11130a8c2cf6c02a09eb ([`jamesob/ackr/23280.5.dongcarl.init_coalesce_chainstate`](https://github.com/jamesob/bitcoin/tree/ackr/23280.5.dongcarl.init_coalesce_chainstate))

Examined interdiff, built, merged & tested recent master and my init functional tests. Changes are as Carl describes.

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK fdc07a1d2916a85add9b11130a8c2cf6c02a09eb ([`jamesob/ackr/23280.5.dongcarl.init_coalesce_chainstate`](https://github.com/jamesob/bitcoin/tree/ackr/23280.5.dongcarl.init_coalesce_chainstate))

Examined interdiff, built, merged & tested recent master and my init functional tests. Changes are as Carl describes.
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmGmduoACgkQepNdrbLE
TwUqWRAAmizHsXiSXi+daWs5gX85JJUb/uetyo2r2qoNFsJh2YDCkrYj3VS/EM5D
JzyscNQKRWSKEhGH7dCP8B1SwK8/7xpcTxw2orOw5Zy4QrbQoZzUy3zKZiV7mrwq
eiWeUlHq2qdruayNIwuWK0JsCDLkPc24jTmsi9Jr+OtvhHDVqfVWz5hRMzP590JX
lDpi92cp7EauuLvrort41uir71WXt7id/n1lusqMRwkvjYDhgAFNG3PS8ooO+kNy
bJRObP55GOizEWoIQ08mIrkbt0wq2ALiQEUFmTvg0ZCFFfuMk4brRrQFoYQwyqc0
RfQrIfaTblLOefRycdDY+wbmNhEhS0ATmeBFrbikzzEcm5Zd9vaV4IaoD5cBsy5e
iKA8mQPX5IW1QC0hHxIEm7bJmJ+5dpwJbqxjm9KzVekw/nR/OjHpPIWj5rGG6xFV
z7Iub6XSjY6w9h/3WSf3hEyTeCnq3kZ5aCUe40tOps1h4jXho70Qtu24l3zA9oza
95dozmg6G2W6JyGl7nPGLoZE27++z2nkBkScs6zOQL8f16uCvXT5FIIM7xt9/uKg
voCeCFE4b3dfSSC4Cce1fx4bt4oGIFfhz6V4nlHcXJdRxaaAs+wcepD8uIR+gdZZ
6Q0oWUrYIQHAi/+ZUqDax2YvNcuZJ6S9fewuS2tQIt+BZG88pMc=
=N1CN
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-5.10.0-9-amd64-x86_64-with-glibc2.28

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis --enable-wallet --enable-debug --with-daemon --enable-natpmp-default CXX=/usr/bin/clang++ CC=/usr/bin/clang PYTHONPATH= --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --disable-openssl-tests --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --no-create --no-recursion

Compiled with /usr/bin/ccache /usr/bin/clang++ -std=c++17 -mavx -mavx2 -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_top_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4 -msha -msse4.1 -msse4.2  i

Compiler version: Debian clang version 13.0.1-++20211110052940+162f3f18c945-1~exp1~20211110053502.25
```

</p></details>

MarcoFalke commented:
`C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\src\node\caches.cpp(32,9): error C7555: use of designated initializers requires at least '/std:c++20' [C:\Users\ContainerAdministrator\AppData\Local\Temp\cirrus-ci-build\build_msvc\libbitcoin_server\libbitcoin_server.vcxproj]`

dongcarl commented:
Did not realize designated initializers is a GNU extension, my bad.

ryanofsky commented on src/init.cpp:
```diff
@@ -1408,9 +1408,14 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
-                                 args.GetIntArg("-checkblocks", DEFAULT_CHECKBLOCKS),
-                                 args.GetIntArg("-checklevel", DEFAULT_CHECKLEVEL));
+                                 nCoinCacheUsage);
+        auto rv2 = VerifyLoadedChainstate(chainman,
```
In commit "Split off VerifyLoadedChainstate" (019da3dc8493054b456222ef23761dd95130b6a7)

Seems like this only be called if `rv` is `nullopt`, and it should be skipped if `LoadChainstate` fails or is interrupted for shutdown. Otherwise this is running verification code that did not run previously.

Would be better to preserve behavior I think. Either way it would also be good to say in commit description whether behavior is changing or not.

ryanofsky commented on src/init.cpp:
```diff
@@ -1408,7 +1408,14 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+                                         "", CClientUIInterface::MSG_ERROR);
+                                 });
+
+        uiInterface.InitMessage(_("Verifying blocks…").translated);
```
In commit "node/chainstate: Decouple from concept of uiInterface" (efe9bd287656925a3bda50a809edeabe9824e37a)

Seems like there is a minor change in behavior with the verifying blocks message being shown even if is_coinsview_empty is true. This seems ok, but would be good to mention whether behavior change is intentional in commit message.

ryanofsky reviewed (commented):
Mostly reviewed (will update list below with progress).
- [X] 6345e65d3163f4f5b4545181813b8e8a5ea5a157 node: Extract chainstate loading sequence (1/22)
- [X] 165b30b649923da4ed19749a399c4a4589da326b node/chainstate: Decouple from GetTimeMillis (2/22)
- [X] 64195bf2e4b8c6f74d0df7ba85a1ee0f0b6c4e5f node/chainstate: Decouple from stringy errors (3/22)
- [X] bb8245bedc18125309ca85e95dfc91eb7c4201bf node/chainstate: Decouple from ArgsManager (4/22)
- [X] aa6bc732b148e5e27bfc1df23f42c6c55006185c node/chainstate: Decouple from concept of NodeContext (5/22)
- [X] 0d2f1a458e5ad8cfbbdc039084ce554daaf41ba9 Move mempool nullptr Assert out of LoadChainstate (6/22)
- [X] 49d2d11aa5dbe13cfcbb11605b2f60e62f4e848b Move init logistics message for BAD_GENESIS_BLOCK to init.cpp (7/22)
- [X] 0b340cd58693d5c4956dbf1eda8f81cf0944b02b node/chainstate: Remove do/while loop (8/22)
- [X] 019da3dc8493054b456222ef23761dd95130b6a7 Split off VerifyLoadedChainstate (9/22)
- [X] efe9bd287656925a3bda50a809edeabe9824e37a node/chainstate: Decouple from concept of uiInterface (10/22)
- [X] 71c78844ee2449e2e8086857bbef57d436cd7b24 node/chainstate: Reduce coupling of LogPrintf (11/22)
- [X] 02c5d48532e9eadfd4832585e827aea53af7aaf5 Move -checkblocks LogPrintf to AppInitMain (12/22)
- [ ] ca13555758e6d868d3cfa5b55e679fca74381b03 init: Delay RPC block notif until warmup finished (13/22)
- [X] e073634c37f3a1e140920c6e5e3f2c1ae47cd293 node/chainstate: Decouple from GetAdjustedTime (14/22)
- [X] 2a2a496fe898404876f4ed12cd61aae8600f3e47 node/chainstate: Decouple from ShutdownRequested (15/22)
- [X] 7d0cf02654214e816734dffe30a466d4d12147f1 validation: VerifyDB only needs Consensus::Params (16/22)
- [ ] fefd26434cc3bcac443a818903df1c31f40ecff1 node/caches: Extract cache calculation logic (17/22)
- [ ] 4a50ef72d7367c35c52ca8f8e584b5f402ae79df node/chainstate: Add options for in-memory DBs (18/22)
- [ ] 526b218273ead38e5d93198367542c4e97956fd8 test/setup: Use LoadChainstate (19/22)
- [ ] 8b3080f2d662207ee0af2aa48a0a6c3db3a24c56 test/setup: Unify m_args and gArgs (20/22)
- [ ] 97162b171e064ebf745fdee1dd3373c38b7195ed Remove all #include // for * comments (21/22)
- [ ] fdc07a1d2916a85add9b11130a8c2cf6c02a09eb Collapse the 2 cs_main locks in LoadChainstate (22/22)

MarcoFalke commented:
```
�[0;36m test  2021-11-30T22:50:44.631000Z TestFramework.node0 (DEBUG): [node 0] bitcoind exited with status -11 during initialization �[0m
�[0;36m test  2021-11-30T22:50:44.632000Z TestFramework (ERROR): Assertion failed �[0m
�[0;36m                                   Traceback (most recent call last):�[0m
�[0;36m                                     File "/private/var/folders/tn/f_9sf1xx5t14qm_6f83q3b840000gn/T/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-apple-darwin19/test/functional/test_framework/test_framework.py", line 132, in main�[0m
�[0;36m                                       self.run_test()�[0m
�[0;36m                                     File "/private/var/folders/tn/f_9sf1xx5t14qm_6f83q3b840000gn/T/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-apple-darwin19/test/functional/feature_init.py", line 168, in run_test�[0m
�[0;36m                                       node.assert_start_raises_init_error(�[0m
�[0;36m                                     File "/private/var/folders/tn/f_9sf1xx5t14qm_6f83q3b840000gn/T/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-apple-darwin19/test/functional/test_framework/test_node.py", line 523, in assert_start_raises_init_error�[0m
�[0;36m                                       self._raise_assertion_error(�[0m
�[0;36m                                     File "/private/var/folders/tn/f_9sf1xx5t14qm_6f83q3b840000gn/T/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-apple-darwin19/test/functional/test_framework/test_node.py", line 167, in _raise_assertion_error�[0m
�[0;36m                                       raise AssertionError(self._node_msg(msg))�[0m
�[0;36m                                   AssertionError: [node 0] Expected message "Error opening block database." does not partially match stderr:�[0m
�[0;36m                                   ""�[0m

jamesob commented:
> ```
> �[0;36m test  2021-11-30T22:50:44.631000Z TestFramework.node0 (DEBUG): [node 0] bitcoind exited with status -11 during initialization �[0m
> �[0;36m test  2021-11-30T22:50:44.632000Z TestFramework (ERROR): Assertion failed �[0m
> �[0;36m                                   Traceback (most recent call last):�[0m
> �[0;36m                                     File "/private/var/folders/tn/f_9sf1xx5t14qm_6f83q3b840000gn/T/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-apple-darwin19/test/functional/test_framework/test_framework.py", line 132, in main�[0m
> �[0;36m                                       self.run_test()�[0m
> �[0;36m                                     File "/private/var/folders/tn/f_9sf1xx5t14qm_6f83q3b840000gn/T/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-apple-darwin19/test/functional/feature_init.py", line 168, in run_test�[0m
> �[0;36m                                       node.assert_start_raises_init_error(�[0m
> �[0;36m                                     File "/private/var/folders/tn/f_9sf1xx5t14qm_6f83q3b840000gn/T/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-apple-darwin19/test/functional/test_framework/test_node.py", line 523, in assert_start_raises_init_error�[0m
> �[0;36m                                       self._raise_assertion_error(�[0m
> �[0;36m                                     File "/private/var/folders/tn/f_9sf1xx5t14qm_6f83q3b840000gn/T/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-apple-darwin19/test/functional/test_framework/test_node.py", line 167, in _raise_assertion_error�[0m
> �[0;36m                                       raise AssertionError(self._node_msg(msg))�[0m
> �[0;36m                                   AssertionError: [node 0] Expected message "Error opening block database." does not partially match stderr:�[0m
> �[0;36m                                   ""�[0m
> ```

Ah yeah, see my comment above (https://github.com/bitcoin/bitcoin/pull/23280#issuecomment-972365657). This should be a simple matter of tacking on a commit to this branch that updates output expectations in the functional test.

ryanofsky commented on src/test/util/setup_common.cpp:
```diff
@@ -77,9 +77,9 @@ std::ostream& operator<<(std::ostream& os, const uint256& num)
 BasicTestingSetup::BasicTestingSetup(const std::string& chainName, const std::vector<const char*>& extra_args)
     : m_path_root{fs::temp_directory_path() / "test_common_" PACKAGE_NAME / g_insecure_rand_ctx_temp_path.rand256().ToString()},
-      m_args{}
+      m_args{gArgs}
```
In commit "test/setup: Unify m_args and gArgs" (7ede150299ea46d38a0797910c95028e7e948e0d)

I think it would be better to drop this change, and let `BasicTestingSetup::m_args` keep being a variable instead of a `gArgs` reference. The reason the `BasicTestingSetup::m_args` member exists is to provide a way for tests to stop using a shared `gArgs` instance and let each test have it's own standalone ArgsManager. This way tests can do `m_args.ForceSetArg` and other operations without affecting global state and potentially breaking each other.

The other change below which stops using `gArgs` would be good to keep if it still works:

```diff
-    m_node.args = &gArgs;
+    m_node.args = &m_args;
```

If it doesn't work now, it should work later when @kiminuo gets around to removing more `gArgs` uses in other code https://github.com/bitcoin/bitcoin/pull/21244#discussion_r610080299

ryanofsky reviewed (approved):
Code review ACK 8301c696d0a0be921ce01b6e9f74f7ff6e8c0f2f. PR is a nice improvement. A few things could be tweaked still, but I don't think any changes are actually needed, and this looks good in its current form.

ryanofsky commented on src/node/chainstate.cpp:
```diff
@@ -18,6 +17,7 @@ std::optional<ChainstateLoadingError> LoadChainstate(bool fReset,
                                                      int64_t nBlockTreeDBCache,
                                                      int64_t nCoinDBCache,
                                                      int64_t nCoinCacheUsage,
+                                                     std::optional<std::function<bool()>> shutdown_requested,
```
In commit "node/chainstate: Decouple from ShutdownRequested" (258ce97cf72043ffa69d2b77a4438df723c734a6)

Function objects can be null so here and other places should replace `optional<function<X>>` with `function<X>` so there is one null value instead of two

ryanofsky commented:
Following up on the return codes discussion https://github.com/bitcoin/bitcoin/pull/23280#discussion_r748465779, I still think it would be good to reduce the number of status codes the caller needs to handle, and number of function parameters. Experimenting a little, I pushed a change (compiles, but untested) to do this here: b7c7f64dc86f0fcdf07cb1e765f7cccc3a3c8897. It could be a followup, or feel free to adopt parts of this if they seem useful.

jamesob reviewed (approved):
reACK 8301c696d0a0be921ce01b6e9f74f7ff6e8c0f2f ([`jamesob/ackr/23280.6.dongcarl.init_coalesce_chainstate`](https://github.com/jamesob/bitcoin/tree/ackr/23280.6.dongcarl.init_coalesce_chainstate))

Changes since my last ACK include
- minor formatting for the `LoadChainstate` call
- break if an error is encountered during `LoadChainstate` instead of proceeding to `VerifyLoadedChainstate` (and displaying the "Verifying blocks..." message)
- don't prefix parameter names when returning from CalculateCacheSizes

I'm still singing the same tune:  this is a big improvement to init, importantly because it can be reused in unittests.

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

reACK 8301c696d0a0be921ce01b6e9f74f7ff6e8c0f2f ([`jamesob/ackr/23280.6.dongcarl.init_coalesce_chainstate`](https://github.com/jamesob/bitcoin/tree/ackr/23280.6.dongcarl.init_coalesce_chainstate))

Changes since my last ACK include
- - minor formatting changes for the `LoadChainstate` call
- - break if an error is encountered during `LoadChainstate` instead of proceeding to `VerifyLoadedChainstate` (and displaying the "Verifying blocks..." message)
- - don't prefix parameter names when returning from CalculateCacheSizes

I'm still singing the same tune:  this is a big improvement to init, importantly because it can be reused in unittests.
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmGqTYkACgkQepNdrbLE
TwWFAg//Suzk9OqS3OCmxCijJEnCWXSv/afG26X5yTqTL/o1DyzQZrFddliL3VJN
Fx6j3Zln80cq1Jl4yeVRsQ8zU7MWv+hpVEyf4SAjcToJ3qxHan3pZfUalPE8nzFv
1ubvztUag/xhO09/W6yBIFGl7IbUDA4uyayg+7QCM2Plv5IXiP3c01ufiGTpuRxg
1H/VszdnQEjBaRLw1r+cv9pGYBUtzgXUk/i17esfAlZV71C/gUOvfVOwp3wHJB2G
PGnpwQ7fQ+cQ/4KED/RIuMMQAztHjPqQUcsOFCFWohtGU7KXbQdD3TjlJYRJNIEW
k5kW+ySVSypJyxuGJVK2wm4Fvu5uzKK2c2tqg714HHMtHxjIz857WTagKpyseKPb
hhWZOSYdQxyinqXtCEF3r1ufBqxjeEPSjAwl2aDOi0bbZ01rBEFAp2VQzfwBSwoz
L81/wLLlQg4AG8LRIcDdIi+yz1xHrBbHbvmjN1Yg8X8fQ7qzmnwxqB9MYz74IE1v
7J4pgxgM6wT772fJIZnlNynVR5ZvlX3VcYTpvos4A4KM7BA2n0KFm3AvQB5s0gUi
omtsbQbysQ5+rqpwUDmhqIhkkSI4pmExmn60Y0UkiLOMrbdKE0AbuGgdQ/p2RvOs
bktDxlmOwraljitKSJHnD0UQuLmdSxuujFW4IO39dN+RrRwsYMQ=
=0Nt0
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-5.10.0-9-amd64-x86_64-with-glibc2.28

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis --enable-wallet --enable-debug --with-daemon --enable-natpmp-default CXX=/usr/bin/clang++ CC=/usr/bin/clang PYTHONPATH= --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --disable-openssl-tests --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --disable-openssl-tests --no-create --no-recursion

Compiled with /usr/bin/ccache /usr/bin/clang++ -std=c++17 -mavx -mavx2 -mpclmul -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_top_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4 -msha -msse4.1 -msse4.2  i

Compiler version: Debian clang version 13.0.1-++20211110052940+162f3f18c945-1~exp1~20211110053502.25
```

</p></details>

jamesob commented:
Needs rebase, though.

jonatack commented:
Approach ACK. Debug build clean, needs rebase.

dongcarl committed:
node/caches: Remove intermediate variables

RandyMcMillan commented:
```
Darwin ₿ 20.5.0 Darwin Kernel Version 20.5.0: Sat May  8 05:10:31 PDT 2021; root:xnu-7195.121.3~9/RELEASE_ARM64_T8101 arm64

```

![Screen Shot 2021-12-06 at 5 32 34 PM](https://user-images.githubusercontent.com/152159/144936209-1fb98b7b-1506-48d9-b18e-c73281fbedc5.png)

![Screen Shot 2021-12-06 at 5 27 26 PM](https://user-images.githubusercontent.com/152159/144936292-2475f52d-6afd-4468-9ed4-611386f63150.png)

mzumsande commented:
@RandyMcMillan: That happens on current master too when building without bdb, see #23684 -  so it's very likely not related to this PR.

jamesob reviewed (approved):
reACK c6861476f7e58f7b6b2e5416d2a7b774d1796a0e ([`jamesob/ackr/23280.7.dongcarl.init_coalesce_chainstate`](https://github.com/jamesob/bitcoin/tree/ackr/23280.7.dongcarl.init_coalesce_chainstate))

Only changes include rebasing on recent master (GetAdjustedTime -> GetTime). Built locally and booted a few times with existing mainnet datadir. 

<details><summary>Show signature data</summary>
<p>

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

reACK c6861476f7e58f7b6b2e5416d2a7b774d1796a0e ([`jamesob/ackr/23280.7.dongcarl.init_coalesce_chainstate`](https://github.com/jamesob/bitcoin/tree/ackr/23280.7.dongcarl.init_coalesce_chainstate))

Only changes include rebasing on recent master (GetAdjustedTime -> GetTime). Built locally and booted a few times with existing mainnet datadir. 
-----BEGIN PGP SIGNATURE-----

iQIzBAEBCgAdFiEEGNRVI1NPYuZCSIrGepNdrbLETwUFAmGvj4IACgkQepNdrbLE
TwW1RBAAnzie/0j3f5h7kn6pH5WLtr4np3ffokobdyfDuPsfhMIp1aqBDX0GQlrs
sy62TEzoPgcD3EQL4TA29Em0/C+uWQIHuBjwnZXSQvFE38BIh4DfcKS+GE2v1j4H
rBp7R60pRXjXNPYv7aTGvhkM2VPsOLt/n4xdSzvzBPjXWmSaCb+rJcivEZHciKXX
4VH2djGJiFwmfmXPo0V9UhNLI86HJ3ju+vJTm32bURTuu7+Hm3P6rHotkhNYGgd7
ZwI35VobxCmU8qd2AD/6H0GwYYr6cketmVi6fwXWpGhzibGvAr7H4nznnNy66AWh
JoVxIPcbCxJMm8WxzD8+Jn5kVMUeX1c1b7oDUGA6o5r3RLJ7tMD0wg6TXMBi2QQ8
E8tC3tTF2NC9cgTBv3xQr1hxx6KIE7oT4tShl6G+ryMtq/kNKCAu2nd1qkEzsQW2
pIcpKUkXBhFrFagUfOri58tKBYEFmohalqjHiEmGAOqqX0FJR7xTEBFGQnlMDUmd
v7/o6swc+w2dp5T7tDtdCtlNxTG57dPndZKEMzlnue6yBtjV8ZPm7PbqSSWhiLV2
ZvNDZp1kH1zbeePRf+bzGU8s448RMEG9c77xyNluTJkpCq+4kRFR1aVreWgJvTh+
OM45ymO4vvYa+0HWqxQUA6CPW/gMr0gkeulQb+MOv/IjpH6dZ7s=
=uM20
-----END PGP SIGNATURE-----

```

</p></details>

<details><summary>Show platform data</summary>
<p>

```
Tested on Linux-5.10.0-9-amd64-x86_64-with-glibc2.28

Configured with ./configure LDFLAGS=-L/home/james/src/bitcoin/db4/lib/ CPPFLAGS=-I/home/james/src/bitcoin/db4/include/ CXXFLAGS=-fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis --enable-wallet --enable-debug --with-daemon --enable-natpmp-default CXX=/usr/bin/clang++ CC=/usr/bin/clang PYTHONPATH= --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --disable-openssl-tests --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-experimental --disable-openssl-tests --no-create --no-recursion

Compiled with /usr/bin/ccache /usr/bin/clang++ -std=c++17 -mavx -mavx2 -mpclmul -fPIE -pipe -O2 -g -Wthread-safety-analysis -Wall -Werror=sign-compare -Wsign-compare -Werror=thread-safety-analysis -O0 -g3 -ftrapv -fdebug-prefix-map=$(abs_top_srcdir)=.  -Wstack-protector -fstack-protector-all -fcf-protection=full -fstack-clash-protection -msse4 -msha -msse4.1 -msse4.2  i

Compiler version: Debian clang version 13.0.1-++20211110052940+162f3f18c945-1~exp1~20211110053502.25
```

</p></details>

dongcarl committed:
style-only: Remove redundant scope in *Chainstate

I strongly recommend reviewing with the following git-diff flags:
  --ignore-space-change

ryanofsky reviewed (approved):
Code review ACK c6861476f7e58f7b6b2e5416d2a7b774d1796a0e. Only change since last review is rebasing after #23636

Maybe this PR is ready for merge. I would still like to drop commit 
"test/setup: Unify m_args and gArgs" (4e54d903a3b4643cc6d91c899a0ac401fc8f6529) which uses `gArgs` more in tests, instead off less (https://github.com/bitcoin/bitcoin/pull/23280#discussion_r760574720)

I also think argument and error code code handling here could be simplified, suggested previously in b7c7f64dc86f0fcdf07cb1e765f7cccc3a3c8897 (+129/-210) https://github.com/bitcoin/bitcoin/pull/23280#issuecomment-984149564

And there might be some other previous suggestions from reviewers that could use followup, but overall PR is an improvement in its current form, already

dongcarl commented:
Pushed c6861476f7e58f7b6b2e5416d2a7b774d1796a0e -> 7f15eff2ddd86034e84a19413e1a42883987f660
- Use `std::function`'s inherent nullability (https://github.com/bitcoin/bitcoin/pull/23280#discussion_r760598661)
- Added code comment in commit "init: Delay RPC block notif until warmup finished" (https://github.com/bitcoin/bitcoin/pull/23280#discussion_r760525288)
- Added commit removing intermediate variables in `CalculateCacheSizes` (https://github.com/bitcoin/bitcoin/pull/23280#discussion_r760560829)
- Removed commit "test/setup: Unify m_args and gArgs" (https://github.com/bitcoin/bitcoin/pull/23280#discussion_r760574720)
- Added commit "style-only: Remove redundant scope in *Chainstate" (https://github.com/bitcoin/bitcoin/pull/23280#discussion_r762078294, https://github.com/bitcoin/bitcoin/pull/23280#discussion_r762079030)

dongcarl commented:
> I also think argument and error code code handling here could be simplified, suggested previously in [b7c7f64](https://github.com/bitcoin/bitcoin/commit/b7c7f64dc86f0fcdf07cb1e765f7cccc3a3c8897) (+129/-210) [#23280 (comment)](https://github.com/bitcoin/bitcoin/pull/23280#issuecomment-984149564)

I think we could put that in a followup PR for discussion, perhaps along with https://github.com/bitcoin/bitcoin/pull/23280#discussion_r762083060

ryanofsky reviewed (approved):
Code review ACK 7f15eff2ddd86034e84a19413e1a42883987f660. Thanks for updates!

laanwj commented on src/init.cpp:
```diff
@@ -1415,6 +1415,8 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
                                 cache_sizes.block_tree_db,
                                 cache_sizes.coins_db,
                                 cache_sizes.coins,
+                                false,
```
Please use named argument style here

ryanofsky replied:
> Please use named argument style here

This suggestion makes sense. But one way or the other I would also like to incorporate or follow up with b7c7f64dc86f0fcdf07cb1e765f7cccc3a3c8897 (+129/-210) [#23280 (comment)](https://github.com/bitcoin/bitcoin/pull/23280#issuecomment-984149564) which gets rid of these arguments.

laanwj replied:
Sure, getting rid of them is even better.

laanwj commented on src/node/chainstate.cpp:
```diff
@@ -143,7 +143,7 @@ std::optional<ChainstateLoadVerifyError> VerifyLoadedChainstate(ChainstateManage
             if (!is_coinsview_empty(chainstate)) {
                 const CBlockIndex* tip = chainstate->m_chain.Tip();
-                if (tip && tip->nTime > GetTime() + MAX_FUTURE_BLOCK_TIME) {
+                if (tip && tip->nTime > get_unix_time_seconds() + 2 * 60 * 60) {
```
As there is no need for precision here, wouldn't it be simpler to pass in the time here instead of a function that returns the time?

MarcoFalke replied:
why is this removing `MAX_FUTURE_BLOCK_TIME`?

This reverts https://github.com/bitcoin/bitcoin/commit/26a1147ce56083d7aa820ac115c16b01e47d911c

MarcoFalke replied:
Also, agree with laanwj. This only calls `is_coinsview_empty`, which shouldn't take more than a second (aka the precision of our clock), so it might be better to just pass in the time.

laanwj commented:
Code review ACK
This is a big refactor but I think it moves things in the right direction:
- Some welcome decoupling, only passing things to functions which they need is great
- Much better to use error codes than string-y errors
- Nice to have `CalculateCacheSizes` return a structure rather than the usual way with tons of output argument

jamesob commented:
I'll reACK after the `MAX_FUTURE_BLOCK_TIME` rebase reversion is fixed

MarcoFalke commented on src/node/chainstate.h:
```diff
@@ -45,10 +45,10 @@ enum class ChainstateLoadingError {
  *        arbitrary points, one of those calls returned true". Therefore, a
  *        return value other than SHUTDOWN_PROBED does not guarantee that
- *        ShutdownRequested() hasn't been called indirectly.
+ *        shutdown_requested() hasn't been called indirectly.
```
I think this comment is "wrong". `shutdown_requested` is a local symbol to the function, so there is no other place calling it.

Maybe just rephrase as "shutdown hasn't been requested indirectly"

MarcoFalke commented on src/node/caches.h:
```diff
@@ -0,0 +1,22 @@
+    int64_t tx_index;
+    int64_t filter_index;
+};
+CacheSizes CalculateCacheSizes(const ArgsManager& args, size_t n_indexes = 0);
```
seems fragile to provide a default value that is unused, no?

maflcko reviewed (approved):
review ACK 7f15eff2ddd86034e84a19413e1a42883987f66 💳

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 7f15eff2ddd86034e84a19413e1a42883987f66 💳
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhRtAv+ICYz6n3kVKJ2bJ1tF9yGZ2fyPthEKAjS/ZUsTPpa2PtQaIhogbJWd3JQ
BsSdTafJk6P9+ODsSyuzwBYZ1NhBwSDS+PtWjvTOV0i5qE04v3Axd8naRWi2GEj+
cai53kzt8vteJdLCq4clcLUsGwfqm1GtWFbZo0gzpZUwBBzR0a4d8w3cEZx8GTL6
S9kvXOKxvnD07B1s+3BXC9bCksJvpAmAAg5U+mU2X0nKOTvlVU0EypZiZ0Ho2Adg
NdRFHUAo2ZSaqvL1nccsmNxJIWvDBKowAwoZFOLsGkJ00YjajBGQW6lyfglb3UO3
9YmFJ1h52sMDKJTL+dYIRvqJrAZ5rKaXgT271+UwvGmxe4Tw2pIn+7cAccypVsXr
9MiwMXnoyYwoCa/5qHz6Pb8VOlpbcYAdrQ1KL1DZ64VKZPEgsCCqGjRvVkarn3WD
j/2UFbANtOkUcorc4A9JgWcAskOXeMF+Bs2ApQWbxVuMsSDXwdNfT3C5C+7tMlhY
qrz4e3ru
=BAAV
-----END PGP SIGNATURE-----
```


</details>

mohammadhosein552zarei commented:
10000

fanquake commented on src/test/util/setup_common.cpp:
```diff
@@ -177,15 +181,18 @@ TestingSetup::TestingSetup(const std::string& chainName, const std::vector<const
+                             m_cache_sizes.coins,
+                             true,
+                             true);
+    assert(!rv.has_value());
```
Followed up with in #23736.

MarcoFalke commented:
@dongcarl Are you going to address the feedback or do you want someone else to do it? Either is fine, just let us know.

dongcarl commented:
@MarcoFalke Thanks for the reminder! Opened #23855.
